File: src/Cpp/examples/example07_prompt_codegen_pipeline.cpp
#include <filesystem>
#include <iostream>
#include <stdexcept>
#include <string>

#include "pipeline_services/config/json_loader.hpp"
#include "pipeline_services/core/registry.hpp"
#include "pipeline_services/examples/text_steps.hpp"
#include "pipeline_services/generated/prompt_actions.hpp"

static std::string findPipelineFile(const std::string& pipelineFileName) {
  std::filesystem::path currentDir = std::filesystem::current_path();
  while (true) {
    const std::filesystem::path candidatePath = currentDir / "pipelines" / pipelineFileName;
    if (std::filesystem::exists(candidatePath)) {
      return candidatePath.string();
    }
    const std::filesystem::path parentDir = currentDir.parent_path();
    if (parentDir == currentDir) {
      break;
    }
    currentDir = parentDir;
  }
  throw std::runtime_error("Could not locate pipelines directory from current working directory");
}

int main() {
  const std::string pipelineFile = findPipelineFile("normalize_name.json");

  pipeline_services::core::PipelineRegistry<std::string> registry;
  registry.registerUnary("strip", pipeline_services::examples::strip);
  pipeline_services::generated::registerGeneratedActions(registry);

  pipeline_services::config::PipelineJsonLoader loader;
  auto pipeline = loader.load_file(pipelineFile, registry);
  const std::string outputValue = pipeline.run("  john   SMITH ");
  std::cout << "output=" << outputValue << std::endl;
  return 0;
}

File: src/Rust/examples/example07_prompt_codegen_pipeline.rs
use std::path::PathBuf;

use pipeline_services::examples::text_steps::strip;
use pipeline_services::generated::register_generated_actions;
use pipeline_services::{PipelineJsonLoader, PipelineRegistry};

fn find_pipeline_file(pipeline_file_name: &str) -> PathBuf {
  let mut current_dir = std::env::current_dir().expect("Failed to determine current directory");
  loop {
    let candidate_path = current_dir.join("pipelines").join(pipeline_file_name);
    if candidate_path.exists() {
      return candidate_path;
    }
    if !current_dir.pop() {
      break;
    }
  }
  panic!("Could not locate pipelines directory from current working directory");
}

fn main() {
  let pipeline_file = find_pipeline_file("normalize_name.json");

  let mut registry: PipelineRegistry<String> = PipelineRegistry::new();
  registry.register_unary("strip", strip);
  register_generated_actions(&mut registry);

  let loader = PipelineJsonLoader::new();
  let pipeline = loader
    .load_file(
      pipeline_file.to_str().expect("Invalid pipeline path"),
      &registry,
    )
    .expect("loader failed");
  let result = pipeline.run("  john   SMITH ".to_string());
  println!("output={}", result.context);
}
